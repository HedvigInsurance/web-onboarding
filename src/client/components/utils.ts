import { keyframes } from '@emotion/core'
import styled from '@emotion/styled'

const spin = keyframes`
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
`
export const Spinner = styled.div`
  display: inline-block;
  width: 1em;
  height: 1em;
  flex: 1;
  border: 2px solid currentColor;
  border-top-color: transparent;
  border-radius: 4rem;
  animation: ${spin} 500ms linear infinite;
`
