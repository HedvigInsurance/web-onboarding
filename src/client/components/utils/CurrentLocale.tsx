import { match } from 'matchly'
import React from 'react'
import { matchPath, useLocation } from 'react-router'
import { Locale } from 'data/graphql'
import { LOCALE_PATH_PATTERN } from 'shared/locale'

export const getLocaleIsoCode = (locale: string): Locale => {
  switch (locale) {
    case 'dk':
      return Locale.DaDk
    case 'dk-en':
      return Locale.EnDk
    case 'no':
      return Locale.NbNo
    case 'no-en':
      return Locale.EnNo
    case 'se':
      return Locale.SvSe
    case 'se-en':
      return Locale.EnSe
    default:
      throw new Error(`Illegal locale "${locale}"`)
  }
}

const getLocaleFromPath = (path: string) => {
  const localeMatch = matchPath<WithLocale>(path, {
    path: LOCALE_PATH_PATTERN + '/*',
  })
  return localeMatch?.params?.locale?.toLowerCase() || ''
}

export interface WithLocale {
  locale: string
}
type RenderProp = (props: { currentLocale: string }) => React.ReactNode

export const CurrentLocale: React.ComponentType<{
  children: RenderProp
}> = ({ children }) => {
  const location = useLocation()
  return (
    <>{children({ currentLocale: getLocaleFromPath(location.pathname) })}</>
  )
}

export const useCurrentLocale = () => {
  const location = useLocation()
  return getLocaleFromPath(location.pathname)
}

export enum Market {
  Se = 'SE',
  No = 'NO',
  Dk = 'DK',
}
export const useMarket = (): Market => {
  const currentLocale = useCurrentLocale()

  return match([
    ['se', Market.Se],
    ['se-en', Market.Se],
    ['no', Market.No],
    ['no-en', Market.No],
    ['dk', Market.Dk],
    ['dk-en', Market.Dk],
  ])(currentLocale)!
}

export const getPickedLocaleFromCurrentLocale = (
  currentLocale: string,
): Locale =>
  match<string, Locale>([
    ['sv_SE', Locale.SvSe],
    ['en_SE', Locale.EnSe],
    ['nb_NO', Locale.NbNo],
    ['en_NO', Locale.EnNo],
    ['da_DK', Locale.DaDk],
    ['en_DK', Locale.EnDk],
  ])(getLocaleIsoCode(currentLocale))!
