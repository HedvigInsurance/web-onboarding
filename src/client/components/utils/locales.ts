import { Locale as IsoLocale } from 'data/graphql'

type LocaleLabel = 'se' | 'se-en' | 'no' | 'no-en' | 'dk' | 'dk-en'

type LocaleData = {
  path: LocaleLabel
  isoLocale: IsoLocale
  marketLabel: 'SE' | 'NO' | 'DK'
  htmlLang: 'en' | 'sv' | 'no' | 'da'
  ssn: Ssn
  birthDate?: BirthDate
}

type Ssn = {
  formatExample: string
  formatRegex: RegExp
}

type BirthDate = {
  formatExample: string
  formatRegex: RegExp
}

const ssnFormats = {
  se: /^([1-2][0-9])([0-9]{2})([0,1][0-9])([0-3][0-9])([0-9]{4})$/,
  no: /^([0-3][0-9])([0,1][0-9])([0-9]{2})([0-9]{5})$/,
  dk: /^([0-3][0-9])([0,1][0-9])([0-9]{2})([0-9]{4})$/,
}

const birthDateFormats = {
  no: /^([0-3][0-9])-([0,1][0-9])-([1-2][0-9][0-9]{2})$/,
  dk: /^([0-3][0-9])-([0,1][0-9])-([1-2][0-9][0-9]{2})$/,
}

export const locales: Record<LocaleLabel, LocaleData> = {
  se: {
    path: 'se',
    isoLocale: IsoLocale.SvSe,
    marketLabel: 'SE',
    htmlLang: 'sv',
    ssn: {
      formatExample: 'ÅÅÅÅMMDDXXXX',
      formatRegex: ssnFormats.se,
    },
  },
  'se-en': {
    path: 'se-en',
    isoLocale: IsoLocale.EnSe,
    marketLabel: 'SE',
    htmlLang: 'en',
    ssn: {
      formatExample: 'YYYYMMDDXXXX',
      formatRegex: ssnFormats.se,
    },
  },
  no: {
    path: 'no',
    isoLocale: IsoLocale.NbNo,
    marketLabel: 'NO',
    htmlLang: 'no',
    ssn: {
      formatExample: 'DDMMÅÅXXXXX',
      formatRegex: ssnFormats.no,
    },
    birthDate: {
      formatExample: 'DD-MM-ÅÅÅÅ',
      formatRegex: birthDateFormats.no,
    },
  },
  'no-en': {
    path: 'no-en',
    isoLocale: IsoLocale.EnNo,
    marketLabel: 'NO',
    htmlLang: 'en',
    ssn: {
      formatExample: 'DDMMYYXXXXX',
      formatRegex: ssnFormats.no,
    },
    birthDate: {
      formatExample: 'DD-MM-YYYY',
      formatRegex: birthDateFormats.no,
    },
  },
  dk: {
    path: 'dk',
    isoLocale: IsoLocale.DaDk,
    marketLabel: 'DK',
    htmlLang: 'da',
    ssn: {
      formatExample: 'DDMMÅÅSSSS',
      formatRegex: ssnFormats.dk,
    },
    birthDate: {
      formatExample: 'DD-MM-ÅÅÅÅ',
      formatRegex: birthDateFormats.dk,
    },
  },
  'dk-en': {
    path: 'dk-en',
    isoLocale: IsoLocale.EnDk,
    marketLabel: 'DK',
    htmlLang: 'en',
    ssn: {
      formatExample: 'DDMMYYSSSS',
      formatRegex: ssnFormats.dk,
    },
    birthDate: {
      formatExample: 'DD-MM-YYYY',
      formatRegex: birthDateFormats.dk,
    },
  },
}
