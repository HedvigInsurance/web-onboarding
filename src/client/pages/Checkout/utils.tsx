import { GraphQLError } from 'graphql'
import { LimitCode } from 'api/quoteBundleErrorSelectors'

export const checkIsManualReviewRequired = (errors: GraphQLError[]) => {
  const manualReviewRequiredError = errors.find((error) => {
    return error?.extensions?.body?.errorCode === 'MANUAL_REVIEW_REQUIRED'
  })

  return manualReviewRequiredError !== undefined
}

export const isSsnInvalid = (errors: GraphQLError[]) => {
  const invalidSsnError = errors.find((error) => {
    return error?.extensions?.body?.errorCode === LimitCode.INVALID_SSN
  })

  return invalidSsnError !== undefined
}
