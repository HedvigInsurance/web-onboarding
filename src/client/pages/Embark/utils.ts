export const afterTick = (f: () => void) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      f()
      resolve()
    }, 0)
  })
}
