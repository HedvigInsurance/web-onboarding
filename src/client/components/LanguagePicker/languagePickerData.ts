import { Market, LocalePath } from 'src/client/components/utils/CurrentLocale'

const LANGUAGE_TITLES = {
  english: 'English',
  swedish: 'Svenska',
  norwegian: 'Norsk',
  danish: 'Dansk',
}

const SHORT_LANGUAGE_TITLES = {
  english: 'En',
  swedish: 'Sv',
  norwegian: 'No',
  danish: 'Da',
}

type Options = {
  linkTo: LocalePath
  title: string
  shortTitle: string
}[]

type LanguagePickerData = Record<Market, Options>

export const languagePickerData: LanguagePickerData = {
  SE: [
    {
      linkTo: 'se',
      title: LANGUAGE_TITLES.swedish,
      shortTitle: SHORT_LANGUAGE_TITLES.swedish,
    },
    {
      linkTo: 'se-en',
      title: LANGUAGE_TITLES.english,
      shortTitle: SHORT_LANGUAGE_TITLES.english,
    },
  ],
  NO: [
    {
      linkTo: 'no',
      title: LANGUAGE_TITLES.norwegian,
      shortTitle: SHORT_LANGUAGE_TITLES.norwegian,
    },
    {
      linkTo: 'no-en',
      title: LANGUAGE_TITLES.english,
      shortTitle: SHORT_LANGUAGE_TITLES.english,
    },
  ],
  DK: [
    {
      linkTo: 'dk',
      title: LANGUAGE_TITLES.danish,
      shortTitle: SHORT_LANGUAGE_TITLES.danish,
    },
    {
      linkTo: 'dk-en',
      title: LANGUAGE_TITLES.english,
      shortTitle: SHORT_LANGUAGE_TITLES.english,
    },
  ],
}
