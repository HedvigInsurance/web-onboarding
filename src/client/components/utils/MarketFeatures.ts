import { Market, useMarket } from './CurrentLocale'

export const useCreditCheckInfo = () => {
  const market = useMarket()

  return [Market.No].includes(market)
}
