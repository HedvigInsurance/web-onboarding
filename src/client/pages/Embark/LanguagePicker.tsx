import styled from '@emotion/styled'
import { colorsV2 } from '@hedviginsurance/brand'
import { motion } from 'framer-motion'
import React from 'react'
import { Link } from 'react-router-dom'
import { useMarket } from 'components/utils/CurrentLocale'
import { LanguageIcon } from 'components/icons/LanguageIcon'
import { languagePickerData } from './languagePickerData'

const Wrapper = styled.div`
  position: relative;
`
const Option = styled.div`
  display: flex;
  flex-direction: column;
`

const LanguageDropdownButton = styled.button`
  background: none;
  padding: 0;
  margin: 0;
  border: none;
  cursor: pointer;
  outline: 0;
`

const StyledLink = styled(Link)`
  outline: 0;
  color: ${colorsV2.black};
  text-decoration: none;
  padding: 10px 20px;
  transition: all 250ms;
  font-size: 0.9rem;

  :active {
    background-color: ${colorsV2.lightgray};
  }
`

const Divider = styled.span`
  height: 1px;
  width: 100%;
  background-color: ${colorsV2.lightgray};
`

const Dropdown = styled.div`
  height: 80px;
  background-color: white;
  border-radius: 5px;
  position: absolute;
  right: 0;
  transform: translateY(10px);
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  overflow: hidden;
  z-index: 9999;
  box-shadow: 0 3px 5px rgba(0, 0, 0, 0.1);
`

export const LanguagePicker: React.FC<{ path?: string }> = ({
  path = '/new-member',
}) => {
  const [isOpen, setIsOpen] = React.useState(false)
  const market = useMarket()

  React.useEffect(() => {
    const clickListener = () => {
      setIsOpen(false)
    }
    document.body.addEventListener('click', clickListener)

    return () => document.body.removeEventListener('click', clickListener)
  }, [])

  return (
    <Wrapper>
      <LanguageDropdownButton onClick={() => setIsOpen(!isOpen)}>
        <LanguageIcon />
      </LanguageDropdownButton>
      <motion.div
        initial={{ opacity: 0, y: -20 }}
        variants={{
          open: { opacity: 1, y: 0 },
          closed: {
            opacity: 0,
            y: -20,
          },
        }}
        animate={isOpen ? 'open' : 'closed'}
        style={{
          pointerEvents: isOpen ? 'auto' : 'none',
          userSelect: isOpen ? 'auto' : 'none',
        }}
        transition={{ type: 'spring', stiffness: 400, damping: 100 }}
      >
        <Dropdown>
          {languagePickerData[market].map(({ linkTo, title }, index) => (
            <Option key={linkTo}>
              <StyledLink to={`/${linkTo}${path}`}>{title}</StyledLink>
              {index < languagePickerData[market].length - 1 && <Divider />}
            </Option>
          ))}
        </Dropdown>
      </motion.div>
    </Wrapper>
  )
}
