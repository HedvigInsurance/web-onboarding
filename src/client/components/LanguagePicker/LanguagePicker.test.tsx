import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { renderComponent } from 'test/utils'
import { LanguagePicker } from './LanguagePicker'

jest.mock('../../apolloClient', () => ({}))

test('Swedish language picker renders without ðŸ’¥', () => {
  const { getByText, queryByText, container } = renderComponent(
    <MemoryRouter initialEntries={['/se/new-member']}>
      <LanguagePicker color="black" />
    </MemoryRouter>,
  )

  expect(getByText('Sv')).toBeVisible()
  expect(getByText('En').getAttribute('href')).toBe('/se-en/new-member')
  expect(queryByText('No')).toBeNull()

  expect(container.firstChild).toMatchSnapshot()
})

test('Norwegian language picker renders without ðŸ’¥', () => {
  const { getByText, queryByText, container } = renderComponent(
    <MemoryRouter initialEntries={['/no/new-member']}>
      <LanguagePicker color="black" />
    </MemoryRouter>,
  )

  expect(getByText('No')).toBeVisible()
  expect(getByText('En').getAttribute('href')).toBe('/no-en/new-member')
  expect(queryByText('Sv')).toBeNull()

  expect(container.firstChild).toMatchSnapshot()
})

test('Danish language picker renders without ðŸ’¥', () => {
  const { getByText, queryByText, container } = renderComponent(
    <MemoryRouter initialEntries={['/dk/new-member']}>
      <LanguagePicker color="black" />
    </MemoryRouter>,
  )

  expect(getByText('Da')).toBeVisible()
  expect(getByText('En').getAttribute('href')).toBe('/dk-en/new-member')
  expect(queryByText('Sv')).toBeNull()

  expect(container.firstChild).toMatchSnapshot()
})
