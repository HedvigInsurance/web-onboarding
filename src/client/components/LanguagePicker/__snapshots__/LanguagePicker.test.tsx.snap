// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Danish language picker renders without ðŸ’¥ 1`] = `
<div
  class="css-1prqaxz-Wrapper e1isu63z0"
>
  <div
    class="css-14fi736-ActiveOption e1isu63z1"
    color="black"
  >
    Da
  </div>
  <span
    class="css-1tysqfo-Divider e1isu63z3"
    color="black"
  />
  <a
    class="css-s5uyen-LinkOption e1isu63z2"
    color="black"
    href="/dk-en/new-member"
  >
    En
  </a>
</div>
`;

exports[`Norwegian language picker renders without ðŸ’¥ 1`] = `
<div
  class="css-1prqaxz-Wrapper e1isu63z0"
>
  <div
    class="css-14fi736-ActiveOption e1isu63z1"
    color="black"
  >
    No
  </div>
  <span
    class="css-1tysqfo-Divider e1isu63z3"
    color="black"
  />
  <a
    class="css-s5uyen-LinkOption e1isu63z2"
    color="black"
    href="/no-en/new-member"
  >
    En
  </a>
</div>
`;

exports[`Swedish language picker renders without ðŸ’¥ 1`] = `
<div
  class="css-1prqaxz-Wrapper e1isu63z0"
>
  <div
    class="css-14fi736-ActiveOption e1isu63z1"
    color="black"
  >
    Sv
  </div>
  <span
    class="css-1tysqfo-Divider e1isu63z3"
    color="black"
  />
  <a
    class="css-s5uyen-LinkOption e1isu63z2"
    color="black"
    href="/se-en/new-member"
  >
    En
  </a>
</div>
`;
